api/python/quilt3/data_transfer.py

```diff
@@ -856,6 +861,76 @@ def calculate_sha256(src_list: List[PhysicalKey], sizes: List[int]):
     return _calculate_sha256_internal(src_list, sizes, [None] * len(src_list))
 
 
+async def _calculate_hash_s3_get_parts(s3_client_provider, s3_executor, s3_pending_parts_semaphore, src, size):
+    loop = asyncio._get_running_loop()
+
+    params = dict(Bucket=src.bucket, Key=src.path)
+    if src.version_id is not None:
+        params.update(VersionId=src.version_id)
+    part_size = s3_transfer_config.multipart_chunksize
+    is_multi_part = (
+        size >= s3_transfer_config.multipart_threshold
+        and size > part_size
+    )
+    part_numbers = (
+        range(math.ceil(size / part_size))
+        if is_multi_part else
+        (None,)
+    )
+
+    s3_client = s3_client_provider.find_correct_client(S3Api.GET_OBJECT, src.bucket, params)
+
+    def read_to_queue(part_number, q):
+        logger.debug('%r part %s: download started', src, part_number)
+        if part_number is not None:
+            start = part_number * part_size
+            end = min(start + part_size, size) - 1
```

Comment:
please add a comment on why -1, is `Range=` exclusive, etc.

Reply:
We already have code like this, I'd like to deduplicate it a later PR.